File: Client/Roll10/Services/PocketbaseService.cs
using System.Reactive.Subjects;
using Microsoft.JSInterop;
using System.Text.Json;
using System.Text.Json.Nodes;
using Roll10.Domain.Models;

namespace Roll10.Services
{
    public record AuthOptions (
        string authUrl,
        string codeChallenge,
        string codeChallengeMethod,
        string codeVerifier,
        string name,
        string state
    );

    public class PocketbaseService 
    {
        private readonly IJSRuntime _js;
        private bool _isInitialized;
        private static readonly Dictionary<string, Func<string, Task>> Callbacks = new ();
        public BehaviorSubject<bool> LoginSubject { get; }
        public BehaviorSubject<User?> UserSubject { get; }

        public PocketbaseService(IJSRuntime js)
        {
            _js = js;
            _isInitialized = false;
            LoginSubject = new BehaviorSubject<bool>(false);
            UserSubject = new BehaviorSubject<User?>(null);
        }

        private async Task CheckInitialization()
        {
            if(_isInitialized)
                return;
            await Initialize(ConfigurationService.Url);
            _isInitialized = true;
        } 

        private async Task Initialize(string url)
        {
            await _js.InvokeVoidAsync("initializePB", url);
        }

        private record AuthReply (
            bool usernamePassword,
            bool emailPassword,
            List<AuthOptions> authProviders
        );

        public async Task<List<AuthOptions>> GetAuthOptions()
        {
            await CheckInitialization();
            var results = await _js.InvokeAsync<AuthReply>("listAuth");
            
            return results.authProviders;
        }

        [JSInvokable]
        public static void HandleEvent(string eventMessage)
        {
            JsonNode data = JsonNode.Parse(eventMessage) ?? new JsonObject();
            var index = data.AsObject()?["record"]?.AsObject()?["collectionName"]?.ToString() ?? "";
            if(Callbacks.ContainsKey(index))
            {
                var record = data.AsObject()?["record"]?.ToJsonString() ?? "{}";
                Callbacks[index](record);
            }
        }

        public async Task SubscribeTo(string collection, string pattern, Func<string, Task> callback)
        {
            await CheckInitialization();
            Callbacks[collection] = callback;
            await _js.InvokeVoidAsync("subscribeToStream",collection, pattern);
        }

        public async Task<bool> IsLoginValid()
        {
            await CheckInitialization();
            var isUserValid = await _js.InvokeAsync<bool>("isUserLoggedIn");
            LoginSubject.OnNext(isUserValid);
            return isUserValid;
        }

        public async Task<User?> GetUser()
        {
            await CheckInitialization();
            var data = await _js.InvokeAsync<string>("getLoginInformation");
            var user = JsonSerializer.Deserialize<User>(data);
            return user;
        }
        
        public async Task UpdateUser()
        {
            await CheckInitialization();
            await _js.InvokeVoidAsync("updateAuth");
            var user = await GetUser();
            UserSubject.OnNext(user);
        }

        public async Task<List<T>> GetFullList<T>(string collectionName, string sort = "+created", string filter = "", string expand = "")
        {
            await CheckInitialization();
            return await _js.InvokeAsync<List<T>>("getFullList", collectionName, sort, filter, expand);
        }

        public async Task<bool> CreateItem<T>(string collectionName, T item)
        {
            await CheckInitialization();
            return await _js.InvokeAsync<bool>("createItem", collectionName, item);
        }

        public async Task<bool> UpdateItem(string collectionName, string recordId, object patchParams)
        {
            await CheckInitialization();
            return await _js.InvokeAsync<bool>("patchItem", collectionName, recordId, patchParams);
        }

        public async Task LogOut()
        {
            await CheckInitialization();
            await _js.InvokeVoidAsync("logout");
            UserSubject.OnNext(null);
            LoginSubject.OnNext(false);
        }
    }
}

File: Client/Roll10Angular/src/app/pocket-base.service.ts
import {Injectable} from '@angular/core';
import PocketBase, {Admin, Record} from "pocketbase";
import {BehaviorSubject} from 'rxjs';
import {Url} from 'src/Configuration';
import {ForceCast, Nullable} from 'src/Helpers';
import {IUser} from "../domain/data/User";

export interface IAuthOptions {
  authUrl: string
  codeChallenge: string
  codeChallengeMethod: string
  codeVerifier: string
  name: string
  stat: string
}

@Injectable({
  providedIn: 'root'
})
export class PocketBaseService {

  private pb = new PocketBase(Url);
  public userSubject = new BehaviorSubject<Nullable<IUser>>(null);
  public callbacks: {
    [key: string]: (event: Record) => Promise<void>
  } = {};

  constructor() { }

  public async getAuthOptions()
  {
    return await this.pb.collection('users').listAuthMethods()
  }

  public GetUser():Nullable<IUser>
  {
    let model = this.pb.authStore.model;
    if(model == null)
      return null;

    if(model instanceof Admin)
    {
      //no admin should login
      this.Logout();
      throw new Error("Admin is not allowed to login");
    }

    return ForceCast<IUser>(model);
  }

  public async UpdateUser()
  {
    try
    {
      let newUser = await this.pb.collection("users").authRefresh();
      this.userSubject.next(ForceCast<IUser>(newUser.record));
    }
    catch(e)
    {
      console.error(e);
      this.userSubject.next(null);
    }
  }

  public async GetFullList<T>(collectionName: string, sort = "+created", filter = "", expand = "")
  {
    try
    {
        let list = await this.pb.collection(collectionName).getFullList(200, {
            sort,
            filter,
            expand
        });
        let expandList = (expand || "").split(",");
        return list.map(l => {
          for (let entry of expandList) {
            if (l["expand"][entry]) {
              l[entry] = l["expand"][entry]
            } else {
              l[entry] = [];
            }
          }
          return l;
        }).map(l => ForceCast<T>(l));
    }
    catch(e)
    {
        console.error(e);
        return [];
    }
  }

  public async CreateItem<T extends {}>(collection: string, item: Partial<T>)
  {
    try
    {
      await this.pb.collection(collection).create(item);
      return true;
    }
    catch(e)
    {
      console.error(e);
      return false;
    }
  }

  public async UpdateItem<T>(collection: string, recordId: string, patchParams: Partial<T>)
  {
    try
    {
      await this.pb.collection(collection).update(recordId, patchParams);
      return true;
    }
    catch(e)
    {
      console.error(e);
      return false;
    }
  }

  public Logout()
  {
    this.pb.authStore.clear();
    this.userSubject.next(null);
  }

  public async SubscribeTo(collection: string, pattern: string, callback: (event: Record) => Promise<void>)
  {
    this.callbacks[collection] = callback;
    this.pb.collection(collection).subscribe(pattern, (e) => {
      this.callbacks[collection](e.record);
    });
  }

  public async HandleAuthRedirect()
  {
    const redirectUrl = window.location.origin + '/redirect';

    // parse the query parameters from the redirected url
    const params = (new URL(window.location.toString())).searchParams;

    // load the previously stored provider's data
    const provider = JSON.parse(localStorage.getItem('provider') ?? "{}")

    // compare the redirect's state param and the stored provider's one
    if (provider.state !== params.get('state')) {
        throw "State parameters don't match.";
    }

    // authenticate
    return this.pb.collection('users').authWithOAuth2(
        provider.name,
        params.get('code') ?? "",
        provider.codeVerifier,
        redirectUrl,
        // pass optional user create data
        {
            emailVisibility: false,
        }
    );
  }
}
